TSDB:  add index timestamp range check

weizijun opened the pull request:
- Add a `index.time_series.start_time` and `index.time_series.end_time` setting to time series indices and reject documents with @timestamp outside the range
- Reject documents without an @timestamp

weizijun committed:
add settings

weizijun committed:
add check code

weizijun committed:
add some comment

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

weizijun committed:
add yaml tests

nik9000 reviewed (commented):
Ah. You already did the end time only getting bigger. It's just the down time.

Could you make sure the setting aren't set-able without being in time series mode? I think right now they are behind the feature flag but not behind enabling the mode. I probably didn't mention that when we typed at each other earlier.

weizijun commented:
> Could you make sure the setting aren't set-able without being in time series mode? I think right now they are behind the feature flag but not behind enabling the mode. I probably didn't mention that when we typed at each other earlier.

ok, I will add the time_series mode check

nik9000 commented:
Thanks!

On Sun, Sep 26, 2021, 07:45 weizijun ***@***.***> wrote:

> Could you make sure the setting aren't set-able without being in time
> series mode? I think right now they are behind the feature flag but not
> behind enabling the mode. I probably didn't mention that when we typed at
> each other earlier.
>
> ok, I will add the time_series mode check
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/78291#issuecomment-927291354>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABUXIWO5LTNGPVRDQH2UG3UD4BVNANCNFSM5EWKXGYA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

weizijun committed:
add the time_series mode check

nik9000 commented on server/src/main/java/org/elasticsearch/index/IndexMode.java:
```diff
@@ -40,6 +51,45 @@ void validateWithOtherSettings(Map<Setting<?>, Object> settings) {
+        }
+
+        private void validateTimestamp(DocumentParserContext context) {
+            IndexableField[] fields = context.rootDoc().getFields(TIMESTAMP_FIELD);
```
I wonder if we'd be better off validating this in `DateFieldMapper` somehow. We can get the single-valued-ness validation with `addWithKey`. We could, maybe, have `IndexMode` get a method like:
```
public abstract LongConsumer dateValidator(String date, Resolution resolution);
```

In non-tsdb it'd just `d -> {}` and in tsdb it'd check.

I think it'd be safe to read the bounds from the IndexSettings with the same volatile stuff you have now. In an ideal world we'd update the mapping if the bounds changed and we wouldn't need these extra volatile reads. But I think that's a bigger change.

nik9000 replied:
@imotov do you think it makes sense to check after the fact like we do here or to push something to the mapper?

imotov replied:
As I mentioned before I would like this to be consistent with the rest of data streams. Since in data streams this is configurable, we need to make it configurable, unless we will decide otherwise in our meeting on Thursday.

weizijun replied:
I prefer DataStreamTimestampFieldMapper to be a common TimestampFieldMapper, then time_series can reuse it. After you discuss with folks on data streams and come to a conclusion, then we can discuss how to modify it.

nik9000 commented on server/src/main/java/org/elasticsearch/index/IndexSettings.java:
```diff
@@ -363,6 +363,66 @@ public void validate(IndexMode value, Map<Setting<?>, Object> settings) {
+            @Override
+            public Iterator<Setting<?>> settings() {
+                final List<Setting<?>> settings = List.of(MODE);
+                return settings.iterator();
```
We should probably validate that `START > END`.

nik9000 commented on server/src/main/java/org/elasticsearch/index/IndexMode.java:
```diff
@@ -40,6 +51,45 @@ void validateWithOtherSettings(Map<Setting<?>, Object> settings) {
+                );
+            }
+
+            if (timestamp > endTime) {
```
I *think* we want end time *exclusive* here - so `>=` I think. Right @imotov ?

imotov replied:
@nik9000 that's a good question. We probably want one of them inclusive and another exclusive. If we want this to be consistent with histograms, then start time should be inclusive and end time exclusive - what do you think?

nik9000 commented on server/src/main/java/org/elasticsearch/index/IndexSettings.java:
```diff
@@ -363,6 +363,66 @@ public void validate(IndexMode value, Map<Setting<?>, Object> settings) {
+            }
+        },
+        Property.IndexScope,
+        Property.Dynamic
```
I think we'd be ok if this one was `Property.Final` - I don't think we need to go back in time on this one. It's only the `end_time` that has to go forwards.

nik9000 commented on server/src/test/java/org/elasticsearch/index/IndexSettingsTests.java:
```diff
@@ -545,4 +549,94 @@ public void testCustomDataPathDeprecated() {
     }
+
+    public void testUpdateTimeSeriesTimeRange() {
+        long startTime = Math.min(randomMillisUpToYear9999(), DateUtils.MAX_MILLIS_BEFORE_9999 - 86400000);
```
I tend to do `TimeUnit.DAY.toMillis(1)` just so its super clear how many milliseconds it is. I always have to double check numbers like `86400000`.

nik9000 commented on server/src/main/java/org/elasticsearch/index/IndexSettings.java:
```diff
@@ -363,6 +363,66 @@ public void validate(IndexMode value, Map<Setting<?>, Object> settings) {
+     * in time series mode, the start time of the index, timestamp must larger than start_time
+     * format is epoch_millis
+     */
+    public static final Setting<Long> TIME_SERIES_START_TIME = Setting.longSetting(
```
I think it'd be a nice follow up change to make a `Setting<Date>` that took date in `DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER`. Just ISO8601 and millis since epoch, basically. Maybe a little kinder to work with. That shouldn't be a "for now" thing though.

imotov reviewed (commented):
@nik9000 let's discuss the timestamp issue.

weizijun committed:
improve

weizijun committed:
Revert "improve"

This reverts commit b4a68c3650d6c6a1211952237d0a894b1c4209dd.

weizijun committed:
Revert "Revert "improve""

This reverts commit 63e552cbca89f94c41d7e603379fdbe90041eabe.

weizijun committed:
Merge branch 'upstream/master' into tsdb-deal-timestreamp

* upstream/master: (521 commits)
  Migrate custom role providers to licensed feature (#79127)
  Remove stale AwaitsFix in InternalEngineTests (#79323)
  Fix errors in RefreshListenersTests (#79324)
  Reeable BwC Tests after #79318 (#79320)
  Mute BwC Tests for #79318 (#79319)
  Reenable BwC Tests after #79308 (#79313)
  Disable BwC Tests for #79308 (#79310)
  Adjust BWC for node-level field cap requests (#79301)
  Allow total memory to be overridden (#78750)
  Fix SnapshotBasedIndexRecoveryIT#testRecoveryIsCancelledAfterDeletingTheIndex (#79269)
  Disable BWC tests
  Mute GeoIpDownloaderCliIT.testStartWithNoDatabases (#79299)
  Add alias support to fleet search API (#79285)
  Create a coordinating node level reader for tsdb (#79197)
  Route documents to the correct shards in tsdb (#77731)
  Inject migrate action regardless of allocate action (#79090)
  Migrate to data tiers should always ensure a TIER_PREFERENCE is set (#79100)
  Skip building of BWC distributions when building release artifacts (#79180)
  Default ENFORCE_DEFAULT_TIER_PREFERENCE to true (#79275)
  Deprecation of transient cluster settings (#78794)
  ...

# Conflicts:
#	rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/tsdb/10_settings.yml
#	server/src/main/java/org/elasticsearch/common/settings/IndexScopedSettings.java
#	server/src/main/java/org/elasticsearch/common/settings/Setting.java
#	server/src/main/java/org/elasticsearch/index/IndexMode.java
#	server/src/test/java/org/elasticsearch/index/TimeSeriesModeTests.java

weizijun committed:
spotless

imotov commented:
There are test failing here. I am not sure if you can see them.

```
org.elasticsearch.xpack.ccr.action.TransportResumeFollowActionTests > testDynamicIndexSettingsAreClassified FAILED
    java.lang.AssertionError: setting [index.time_series.end_time] is not classified as replicated or not replicated
    Expected: is <true>
         but: was <false>
        at __randomizedtesting.SeedInfo.seed([9AB92C831026ACBD:2F05BBEA3A1C9FC7]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.elasticsearch.xpack.ccr.action.TransportResumeFollowActionTests.testDynamicIndexSettingsAreClassified(TransportResumeFollowActionTests.java:227)
REPRODUCE WITH: ./gradlew ':x-pack:plugin:ccr:test' --tests "org.elasticsearch.xpack.ccr.action.TransportResumeFollowActionTests.testDynamicIndexSettingsAreClassified" -Dtests.seed=9AB92C831026ACBD -Dtests.locale=ar-EG -Dtests.timezone=Africa/Windhoek -Druntime.java=11

Suite: Test class org.elasticsearch.xpack.ccr.action.TransportResumeFollowActionTests
  1> [2021-10-18T22:07:28,045][INFO ][o.e.x.c.a.TransportResumeFollowActionTests] [testDynamicIndexSettingsAreClassified] before test
  1> [2021-10-18T22:07:28,568][INFO ][o.e.x.c.a.TransportResumeFollowActionTests] [testDynamicIndexSettingsAreClassified] after test
  2> REPRODUCE WITH: ./gradlew ':x-pack:plugin:ccr:test' --tests "org.elasticsearch.xpack.ccr.action.TransportResumeFollowActionTests.testDynamicIndexSettingsAreClassified" -Dtests.seed=9AB92C831026ACBD -Dtests.locale=ar-EG -Dtests.timezone=Africa/Windhoek -Druntime.java=11
  2> java.lang.AssertionError: setting [index.time_series.end_time] is not classified as replicated or not replicated
    Expected: is <true>
         but: was <false>
        at __randomizedtesting.SeedInfo.seed([9AB92C831026ACBD:2F05BBEA3A1C9FC7]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.elasticsearch.xpack.ccr.action.TransportResumeFollowActionTests.testDynamicIndexSettingsAreClassified(TransportResumeFollowActionTests.java:227)
  1> [2021-10-18T22:07:28,617][INFO ][o.e.x.c.a.TransportResumeFollowActionTests] [testValidation] before test
  1> [2021-10-18T22:07:29,608][INFO ][o.e.x.c.a.TransportResumeFollowActionTests] [testValidation] after test
  1> [2021-10-18T22:07:29,631][INFO ][o.e.x.c.a.TransportResumeFollowActionTests] [testFilter] before test
  1> [2021-10-18T22:07:29,632][INFO ][o.e.x.c.a.TransportResumeFollowActionTests] [testFilter] after test
  2> NOTE: leaving temporary files on disk at: /dev/shm/elastic+elasticsearch+pull-request+part-2/x-pack/plugin/ccr/build/testrun/test/temp/org.elasticsearch.xpack.ccr.action.TransportResumeFollowActionTests_9AB92C831026ACBD-001
  2> NOTE: test params are: codec=Asserting(Lucene90): {}, docValues:{}, maxPointsInLeafNode=803, maxMBSortInHeap=7.284619573576025, sim=Asserting(RandomSimilarity(queryNorm=false): {}), locale=ar-EG, timezone=Africa/Windhoek
  2> NOTE: Linux 5.10.0-9-cloud-amd64 amd64/Oracle Corporation 11.0.12 (64-bit)/cpus=32,threads=1,free=429254720,total=536870912
  2> NOTE: All tests run in this JVM: [ActivateAutoFollowPatternActionRequestTests, ShardChangesResponseTests, TransportResumeFollowActionTests]
Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8 -Dsun.jnu.encoding=UTF8
```

```
org.elasticsearch.index.TimeSeriesModeTests > testValidateTimestamp FAILED
    java.lang.IllegalArgumentException: [index.mode=time_series] requires [index.routing_path]
        at __randomizedtesting.SeedInfo.seed([84AD7654F32C4F66:426B8A67AE365071]:0)
        at org.elasticsearch.index.IndexMode$2.validateWithOtherSettings(IndexMode.java:70)
        at org.elasticsearch.index.IndexSettings$1.validate(IndexSettings.java:356)
        at org.elasticsearch.index.IndexSettings$1.validate(IndexSettings.java:350)
        at org.elasticsearch.common.settings.Setting.get(Setting.java:484)
        at org.elasticsearch.common.settings.Setting.get(Setting.java:455)
        at org.elasticsearch.common.settings.Setting.get(Setting.java:598)
        at org.elasticsearch.common.settings.AbstractScopedSettings.get(AbstractScopedSettings.java:709)
        at org.elasticsearch.index.IndexSettings.<init>(IndexSettings.java:581)
        at org.elasticsearch.index.IndexSettings.<init>(IndexSettings.java:561)
        at org.elasticsearch.index.mapper.MapperServiceTestCase.createIndexSettings(MapperServiceTestCase.java:212)
        at org.elasticsearch.index.mapper.MapperServiceTestCase.createMapperService(MapperServiceTestCase.java:183)
        at org.elasticsearch.index.mapper.MapperServiceTestCase.createMapperService(MapperServiceTestCase.java:171)
        at org.elasticsearch.index.mapper.MapperServiceTestCase.createMapperService(MapperServiceTestCase.java:140)
        at org.elasticsearch.index.TimeSeriesModeTests.testValidateTimestamp(TimeSeriesModeTests.java:83)
REPRODUCE WITH: ./gradlew ':server:test' --tests "org.elasticsearch.index.TimeSeriesModeTests.testValidateTimestamp" -Dtests.seed=84AD7654F32C4F66 -Dtests.locale=nb -Dtests.timezone=America/Marigot -Druntime.java=11

```

weizijun committed:
Merge branch 'upstream/master' into tsdb-deal-timestreamp

* upstream/master: (209 commits)
  Enforce license expiration (#79671)
  TSDB: Automatically add timestamp mapper (#79136)
  [DOCS]  `_id` is required for bulk API's `update` action (#79774)
  EQL: Add optional fields and limit joining keys on non-null values only (#79677)
  [DOCS] Document range enrich policy (#79607)
  [DOCS] Fix typos in 8.0 security migration (#79802)
  Allow listing older repositories (#78244)
  [ML] track inference model feature usage per node (#79752)
  Remove IncrementalClusterStateWriter & related code (#79738)
  Reuse previous indices lookup when possible (#79004)
  Reduce merging in PersistedClusterStateService (#79793)
  SQL: Adjust JDBC docs to use milliseconds for timeouts (#79628)
  Remove endpoint for freezing indices (#78918)
  [ML] add timeout parameter for DELETE trained_models API (#79739)
  [ML] wait for .ml-state-write alias to be readable (#79731)
  [Docs] Update edgengram-tokenizer.asciidoc (#79577)
  [Test][Transform] fix UpdateTransformActionRequestTests failure (#79787)
  Limit CS Update Task Description Size (#79443)
  Apply the reader wrapper on can_match source (#78988)
  [DOCS] Adds new transform limitation item and a note to the tutorial (#79479)
  ...

# Conflicts:
#	server/src/main/java/org/elasticsearch/index/IndexMode.java
#	server/src/test/java/org/elasticsearch/index/TimeSeriesModeTests.java

weizijun committed:
improve

weizijun committed:
fix ccr tests

weizijun committed:
Merge branch 'master' into tsdb-deal-timestreamp

weizijun committed:
spotless

weizijun committed:
improve

elasticmachine committed:
Merge branch 'master' into tsdb-deal-timestreamp

imotov commented:
@elasticmachine test this please

weizijun commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into tsdb-deal-timestreamp

weizijun committed:
fix bwc

weizijun commented:
> @elasticmachine test this please

@imotov I fixed the bwc failed tests. The reason for the failure is that the latest version is 8.1.0, and the settings skip version is before 8.0.0. I fixed it.

imotov commented:
@elasticmachine test this please

nik9000 commented on server/src/main/java/org/elasticsearch/index/IndexMode.java:
```diff
@@ -74,6 +81,28 @@ void validateWithOtherSettings(Map<Setting<?>, Object> settings) {
+            if (timestamp >= endTime) {
+                throw new IllegalArgumentException(
+                    "time series index " + TIMESTAMP_FIELD + " value [" + timestamp + "] must be smaller than " + endTime
+                );
```
You pinged me and asked for another review but haven't change this so I'll restate my objections - I feel like, as much as possible, `IndexMode` should be about making sure that the appropriate settings are set and the appropriate mapping is in place. Then we can let the mapping do the assertions.  So I think it's more right to do the test in `DataStreamTimestampFieldMapper` and do it something like:
```
if (context.settings.startTime.isSet) {
  if (time < context.settings.startTime) {
...
```

I like this because a reader doesn't have to know what time series mode is - they just have to know that they is some way to set the min and max time. If they need to figure out how to set it then they can. But lots of times I find myself not caring where a behavior comes from, just that we have this behavior.

weizijun committed:
improve

nik9000 committed:
Maybe like this?

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java:
```diff
@@ -201,6 +201,22 @@ NumericType numericType() {
+            if (value >= endTime) {
+                throw new IllegalArgumentException("time series index @timestamp value [" + value + "] must be smaller than " + endTime);
+            }
+        }
```
I know I wrote this, but on further reading, I think maybe it makes sense to just move it all to `DataStreamTimestampFieldMapper`. I thought I was being clever, but all of this code should be able to run just fine from there and it'd be nice that `DateFieldMapper` doesn't need to care about the validation.

martijnvg replied:
I agree with moving this logic to `DataStreamTimestampFieldMapper`, since there there is already validation for the `@timestamp` field (whether it exists and whether a single value has been specified). And I think at least initially, tsdb will work only with data streams.

weizijun committed:
fix failed tests

weizijun committed:
fix failed tests

nik9000 commented on server/src/main/java/org/elasticsearch/index/IndexSettings.java:
```diff
@@ -509,6 +551,14 @@ public void validate(IndexMode value, Map<Setting<?>, Object> settings) {
+    /**
+     * End time of the time_series index.
+     */
+    private volatile Instant timeSeriesEndTime;
```
I'm worried that right now these can't be `null` so all indices are going to validate the range. I wonder if we can convert these to `null` for the defaults so we disable the validation when it isn't set.

weizijun committed:
fix failed tests

weizijun committed:
improve

weizijun committed:
checkStyle

weizijun committed:
spotless

nik9000 commented:
@elasticmachine, test this please

weizijun committed:
fixex tsdb settings tests

nik9000 commented:
@elasticmachine, test this please

nik9000 commented:
@elasticmachine, test this please

nik9000 committed:
Merge branch 'master' into tsdb-deal-timestreamp

nik9000 commented:
@elasticmachine, test this please

nik9000 committed:
Fixup after merge

nik9000 commented:
> All checks have passed

Nice

nik9000 commented:
Thanks for all that back and for @weizijun ! I've merged master in and merged it.
